Runtime/Tracking/Collision/CollisionTracker.cs

```diff
@@ -1,19 +1,66 @@
 ﻿namespace Zinnia.Tracking.Collision
 {
     using UnityEngine;
+    using System.Linq;
+    using System.Collections.Generic;
+    using Malimbe.MemberClearanceMethod;
+    using Malimbe.XmlDocumentationAttribute;
+    using Malimbe.PropertySerializationAttribute;
+    using Zinnia.Data.Attribute;
 
     /// <summary>
     /// Tracks collisions on the <see cref="GameObject"/> this component is on.
     /// </summary>
     public class CollisionTracker : CollisionNotifier
     {
+        /// <summary>
+        /// Causes collisions to stop if the <see cref="GameObject"/> either side of the collision is disabled.
+        /// </summary>
+        [Serialized]
+        [field: DocumentedByXml, Restricted(RestrictedAttribute.Restrictions.ReadOnlyAtRunTime)]
+        public bool StopCollisionsOnDisable { get; protected set; } = true;
+
+        /// <summary>
+        /// A collection of current existing collisions.
+        /// </summary>
+        protected List<Collider> trackedCollisions = new List<Collider>();
+
+        /// <summary>
+        /// Stops the collision between this <see cref="CollisionTracker"/> and the given <see cref="Collider"/>. If there is still a physical intersection, then the collision will start again next physics frame.
+        /// </summary>
+        /// <param name="collider">The collider to stop the collision with.</param>
+        public virtual void StopCollision(Collider collider)
+        {
+            if ((StatesToProcess & CollisionStates.Exit) == 0)
+            {
+                return;
+            }
+
+            RemoveTrackedNode(collider);
+            OnCollisionStopped(eventData.Set(this, collider.isTrigger, null, collider));
+        }
+
+        protected virtual void OnDisable()
+        {
+            if (!StopCollisionsOnDisable)
+            {
+                return;
+            }
+
+            foreach (Collider collider in trackedCollisions.ToArray())
+            {
+                StopCollision(collider);
+            }
+        }
+
         protected virtual void OnCollisionEnter(Collision collision)
         {
             if ((StatesToProcess & CollisionStates.Enter) == 0)
             {
                 return;
             }
 
+            AddTrackedNode(collision.collider);

```

Comment:
Move these calls (`AddTrackedNode` and `RemoveTrackedNode`) into the `On*` (e.g. this one's `OnCollisionStarted`) methods we call here. I.e. override the methods in this subclass.

Since the `On*` methods are our abstraction it feels like it makes more sense to do this additional logic in them rather than here, just for the sake of ensuring it's done when one would expect it to. This also ensure subclasses don't break that functionality when calling the `On*` methods themselves.

Reply:
Well this fix is actually fixing the issue Unity has with `OnCollisionXXX` and `OnTriggerXXX` what if there was a different implementation that extended the `CollisionNotifier`? You may not even need this work around.

I kinda feel like the `CollisionNotifier` is super generic and the `CollisionTracker` is our implementation of how Unity tracks collisions (badly).

Which would be my case for keeping these only in here and not dirtying the `CollisionNotifier`..... thoughts?

Reply:
Oh that's true. Anyone else using the `On*` methods should have proper ways to work with them right away, and if they don't they most likely are using this subclass anyway, so it wouldn't matter how we do it. As long as it's happening starting from Unity's magic callback methods that is, which this solution does.

Closing as resolved :)